Dask usage at blue yonder

andreas-merkel-by opened the pull request:
This blog post is about how we are using Dask.

dependabot-preview[bot] committed:
Bump github-pages from 197 to 199

Bumps [github-pages](https://github.com/github/pages-gem) from 197 to 199.
- [Release notes](https://github.com/github/pages-gem/releases)
- [Commits](https://github.com/github/pages-gem/compare/v197...v199)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

sebastian-neubauer-by committed:
Add post Write-Up of the PyConDE & PyData Berlin 2019 conference

sebastian-neubauer-by committed:
Remove table of content

sebastian-neubauer-by committed:
Fix typo in heading

sebastian-neubauer-by committed:
Fix repetition

Co-Authored-By: Lucas Rademaker <44430780+lr4d@users.noreply.github.com>

sebastian-neubauer-by committed:
Fix missing comma

Co-Authored-By: Lucas Rademaker <44430780+lr4d@users.noreply.github.com>

sebastian-neubauer-by committed:
Remove bogus "so"

Co-Authored-By: Lucas Rademaker <44430780+lr4d@users.noreply.github.com>

sebastian-neubauer-by committed:
Minor rephrasing

Co-Authored-By: Lucas Rademaker <44430780+lr4d@users.noreply.github.com>

sebastian-neubauer-by committed:
Minor rephrasing

Co-Authored-By: Lucas Rademaker <44430780+lr4d@users.noreply.github.com>

sebastian-neubauer-by committed:
Minor rephrasing

Co-Authored-By: Lucas Rademaker <44430780+lr4d@users.noreply.github.com>

sebastian-neubauer-by committed:
Minor rephrasing

Co-Authored-By: Lucas Rademaker <44430780+lr4d@users.noreply.github.com>

sebastian-neubauer-by committed:
Minor rephrase

sebastian-neubauer-by committed:
Minor rephrasing

Co-Authored-By: Lucas Rademaker <44430780+lr4d@users.noreply.github.com>

lr4d committed:
add lucas rademaker author

lr4d committed:
dask developer workshop blog post

sebastian-neubauer-by committed:
Switched motivation to covid-19

sebastian-neubauer-by committed:
Merge branch 'add_post_pycon_2019_writeup' of https://github.com/JDASoftwareGroup/jdatech.github.io into add_post_pycon_2019_writeup

sebastian-neubauer-by committed:
Minor rephrasing

Co-Authored-By: Lucas Rademaker <44430780+lr4d@users.noreply.github.com>

sebastian-neubauer-by committed:
Minor rephrasing

gabriel-kohen-by committed:
Merge pull request #10 from JDASoftwareGroup/dependabot-bundler-github-pages-199

Bump github-pages from 197 to 199

dependabot-preview[bot] committed:
Bump minimal-mistakes-jekyll from 4.15.2 to 4.16.5

Bumps [minimal-mistakes-jekyll](https://github.com/mmistakes/minimal-mistakes) from 4.15.2 to 4.16.5.
- [Release notes](https://github.com/mmistakes/minimal-mistakes/releases)
- [Changelog](https://github.com/mmistakes/minimal-mistakes/blob/master/CHANGELOG.md)
- [Commits](https://github.com/mmistakes/minimal-mistakes/compare/4.15.2...4.16.5)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
[Security] Bump rubyzip from 1.2.2 to 2.3.0

Bumps [rubyzip](https://github.com/rubyzip/rubyzip) from 1.2.2 to 2.3.0. **This update includes a security fix.**
- [Release notes](https://github.com/rubyzip/rubyzip/releases)
- [Changelog](https://github.com/rubyzip/rubyzip/blob/master/Changelog.md)
- [Commits](https://github.com/rubyzip/rubyzip/compare/v1.2.2...v2.3.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
[Security] Bump nokogiri from 1.10.4 to 1.10.9

Bumps [nokogiri](https://github.com/sparklemotion/nokogiri) from 1.10.4 to 1.10.9. **This update includes security fixes.**
- [Release notes](https://github.com/sparklemotion/nokogiri/releases)
- [Changelog](https://github.com/sparklemotion/nokogiri/blob/v1.10.9/CHANGELOG.md)
- [Commits](https://github.com/sparklemotion/nokogiri/compare/v1.10.4...v1.10.9)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

gabriel-kohen-by committed:
Merge pull request #20 from JDASoftwareGroup/dependabot-bundler-nokogiri-1.10.9

[Security] Bump nokogiri from 1.10.4 to 1.10.9

gabriel-kohen-by committed:
Merge pull request #22 from JDASoftwareGroup/dependabot-bundler-rubyzip-2.3.0

[Security] Bump rubyzip from 1.2.2 to 2.3.0

gabriel-kohen-by committed:
Merge pull request #12 from JDASoftwareGroup/dependabot-bundler-minimal-mistakes-jekyll-4.16.5

Bump minimal-mistakes-jekyll from 4.15.2 to 4.16.5

gabriel-kohen-by committed:
Merge pull request #21 from JDASoftwareGroup/add_post_pycon_2019_writeup

Add post Write-Up of the PyConDE & PyData Berlin 2019 conference

lr4d committed:
add Nefta Kanilmaz to authors

lr4d committed:
wip: address review comments

lr4d committed:
wip: comments from florian

lr4d committed:
wip

lr4d committed:
wip

lr4d committed:
add dask image, fix author

lr4d committed:
fix authors

lr4d commented on _posts/2020-03-16-dask-usage-at-blue-yonder.markdown:
```diff
@@ -0,0 +1,159 @@
+  overlay_image: assets/images/tech_gear_banner.jpg
+  overlay_filter: 0.2
+  show_overlay_excerpt: false
+author: Andreas Merkel, Kshitij Mathur
```
You need to have the entry of `Andreas Merkel, Kshitij Mathur` in `authors.yml` for this to "work", e.g. https://github.com/JDASoftwareGroup/jdatech.github.io/blob/1b4ab751c9946321272f4a918fd81db509113af7/_data/authors.yml#L71
(You can test this locally using the last command provided in the readme)

andreas-merkel-by replied:
Thx, I was already suspecting it's not that smart...

lr4d commented:
btw you may want to take a look at https://github.com/JDASoftwareGroup/jdatech.github.io/pull/23, possibly link it, we also go over some dask issues while giving an overview of the Dask Dev workshop
That should be published _soon_

lucas-rademaker-by committed:
grammar/phrasing corrections

Co-authored-by: KrishnarajAjayGharpure <32620212+KrishnarajAjayGharpure@users.noreply.github.com>

lucas-rademaker-by committed:
Update _posts/2020-03-24-dask-developer-workshop.markdown

lr4d committed:
address krish's comments

lr4d committed:
address sebi's comments

lr4d committed:
change date

gabriel-kohen-by committed:
Merge pull request #23 from JDASoftwareGroup/dask_blog_post

Dask blog post

lr4d commented on _posts/2020-03-16-dask-usage-at-blue-yonder.markdown:
```diff
@@ -0,0 +1,161 @@
+[workers are removed from the cluster](https://github.com/dask/distributed/pull/3366). 
+
+If you are interested in our contributions to Dask and our commitment to the dask community, please
+also check out our blog post [Karlsruhe to D.C. ― a Dask story](2020-03-24-dask-developer-workshop.markdown).
```
```suggestion
also check out our blog post [Karlsruhe to D.C. ― a Dask story](2020-05-28-dask-developer-workshop.markdown).
```
We just changed the date before publishing. You may want to rebase and check using the jekyll commands

jeff-edgett-jdas committed:
Update CNAME

gabriel-kohen-by committed:
Merge pull request #26 from JDASoftwareGroup/jeff-edgett-jdas-patch-1

Update CNAME

jeff-edgett-jdas committed:
Update _config.yml

jeff-edgett-jdas committed:
Update navigation.yml

jeff-edgett-jdas committed:
Update README.md

jeff-edgett-jdas committed:
Delete jda_techblog_w.png

jeff-edgett-jdas committed:
Delete jda_techblog_rgb.png

jeff-edgett-jdas committed:
Delete jda_rgb.png

jeff-edgett-jdas committed:
Add new Blue Yonder Tech Blog lockup

jeff-edgett-jdas committed:
Update _config.yml

jeff-edgett-jdas committed:
Create footer.html

jeff-edgett-jdas committed:
Merge pull request #27 from JDASoftwareGroup/jeff-edgett-jdas-patch-1

Initial Pass of Branding Changes

jeff-edgett-jdas committed:
Additional branding changes...

This shuffles some of the footer menu, makes changes to branding colors and font.

jeff-edgett-jdas committed:
Update Twitter Handle

jeff-edgett-jdas committed:
Update main.scss

Adjust font sizes to account for Minimal Mistakes themes relative font sizing.

gabriel-kohen-by committed:
Merge pull request #28 from JDASoftwareGroup/jeff-edgett-jdas-patch-1

Additional branding changes...

jeff-edgett-jdas committed:
Fix legal links in footer, styling changes

Further styling changes to headings, page display, search results

jeff-edgett-jdas committed:
Update font sizes in search results

jeff-edgett-jdas committed:
Hiding Twitter until account name can be resolved.

gabriel-kohen-by committed:
Merge pull request #29 from JDASoftwareGroup/jeff-edgett-jdas-patch-1

Additional branding changes...

frank-fischer-by commented on _posts/2020-03-16-dask-usage-at-blue-yonder.markdown:
```diff
@@ -0,0 +1,161 @@
+
+### Cluster separation
+
+We are running separate Dask clusters for separate customers in order to isolate the customers from each other.
```
What is a separate customer? ;-)

```suggestion
To provide isolatation between customer environments we run separate Dask clusters per customer.
```

frank-fischer-by commented on _posts/2020-03-16-dask-usage-at-blue-yonder.markdown:
```diff
@@ -0,0 +1,161 @@
+
+First, let's have a look at the use cases we have for Dask. The use of Dask at Blue Yonder
+is strongly coupled to the concept of [Datasets](https://tech.jda.com/introducing-kartothek/),
+tabular data stored as Apache [Parquet](https://parquet.apache.org/) files in blob storage. 
```
```suggestion
tabular data stored as [Apache Parquet](https://parquet.apache.org/) files in blob storage. 
```

frank-fischer-by commented on _posts/2020-03-16-dask-usage-at-blue-yonder.markdown:
```diff
@@ -0,0 +1,161 @@
+---
```
I know I'm a markdown Nazi, but I personally _really_ like to rely on markdown's formatter, i.e. one sentence = one line. That's IMHO much easier to review/comment then mixed sentences in one line.

```
This is a sentence.
This is another sentence.
And this is how markdown renders it.
```

This is a sentence. This is another sentence. And this is how Markdown renders it if everything is configured correctly.

lr4d replied:
Why not a paragraph per line? Then, what you see is what you get and it makes inserting suggestions easier

frank-fischer-by replied:
IMHO then it's pretty hard to review.

The changelog might look like 60 lines for adding a single `,`.

marco-neumann-by replied:
Or if you don't like to do sentences, then at least use a reasonable line length (like 120chars).

andreas-merkel-by replied:
I think that's what I did. For longer sentences, the lines get longer than I like otherwise. But it might be a good idea to start each new sentence on a new line.

frank-fischer-by commented on _posts/2020-03-16-dask-usage-at-blue-yonder.markdown:
```diff
@@ -0,0 +1,161 @@
+First, let's have a look at the use cases we have for Dask. The use of Dask at Blue Yonder
+is strongly coupled to the concept of [Datasets](https://tech.jda.com/introducing-kartothek/),
+tabular data stored as Apache [Parquet](https://parquet.apache.org/) files in blob storage. 
+We use these datasets for storing everything that is needed
+for generating predictions, recommended order quantities, and recommended prices for our customers,
+as well as the predictions, order quantities, and prices themselves.
```
I needed to read that multiple times to really understand the sentence.

1. Emphasize `storing`?
2. The enumeration of `for generating predictions, recommended order quantities, and recommended prices` looks wrong. You want to stick to one tense/format, i.e. `recommended` → `recommend` or even further to change the verb.
3. Do you mean _Dask_ and not the datasets here? I still do not understand why we are mentioning them twice... 
```suggestion
We use datasets for storing everything that is needed
for generating predictions and creating order quantity recommendations or price recommendations for our customers.
Dask is used in managing/creating this data as well as performing the necessary computations.
```

frank-fischer-by commented on _posts/2020-03-16-dask-usage-at-blue-yonder.markdown:
```diff
@@ -0,0 +1,161 @@
+for generating predictions, recommended order quantities, and recommended prices for our customers,
+as well as the predictions, order quantities, and prices themselves.
+
+The size of the data we work on varies strongly depending on the individual customer, with the largest ones
+currently amounting to about 300 GiB of compressed Parquet data per day, corresponding to about 500 GiB of
+uncompressed data in memory.
```
```suggestion
The size of the data we work on varies strongly depending on the individual customer.
The largest ones currently amount to about 300 GiB of compressed Parquet data per day, corresponding to about 500 GiB of uncompressed data in memory.
```

frank-fischer-by commented on _posts/2020-03-16-dask-usage-at-blue-yonder.markdown:
```diff
@@ -0,0 +1,161 @@
+
+## Ad hoc data analysis
+
+Finally, our data scientists also use Dask for ad hoc analysis of data that is too large to
+fit into an individual compute node's memory.
```
Should we incorporate the buzzword `out-of-core` here?

fjetter replied:
+1 for buzzword bingo

We can also mention that this happens from within a jupyter notebook (mentioning OSS technology is usually a good idea)

frank-fischer-by commented on _posts/2020-03-16-dask-usage-at-blue-yonder.markdown:
```diff
@@ -0,0 +1,161 @@
+## Dask cluster setup at Blue Yonder
+
+We run a somewhat unique setup of Dask clusters that is partly driven by the specific requirements
+of our domain and partly historically grown. We are running a large number of individual clusters
```
We run ... We are running ...?

```suggestion
of our domain and partly historically grown. We operate a large number of individual clusters
```

fjetter replied:
I would argue this is not historically grown but intentionally. Keyword: "environment isolation for SaaS applications" (And data isolation between customers) .This can be argued but this is how I motivated it when I was asked about it during the dask workshop

frank-fischer-by commented on _posts/2020-03-16-dask-usage-at-blue-yonder.markdown:
```diff
@@ -0,0 +1,161 @@
+We are running separate Dask clusters for separate customers in order to isolate the customers from each other.
+For the same reason, we also run separate clusters for production, staging, and development environments. 
+
+But it does not stop there. Our compute pipeline involves different steps, for instance, downloading the input data
```
```suggestion
But it does not stop there. Our compute pipeline involves different steps, for instance downloading the input data
```

unless there is some comma rule that I don't know about ;-)

fjetter replied:
This section describes what I referred to as environment isolation.
I would motivate it with having different products which build onto each others and are running sequentially. Since every product / application operates mostly autonomously, we separated the environments off each others. We might also drop a note about having different runtime environnments (python only, C++, lately even java, etc.)

frank-fischer-by replied:
But that's the **rest of this paragraph**, that you're talking about.
This sentence/part really focuses on what the pipeline is comprised of and therefore is not really Cluster separation?!


So IMHO, this sections does not convey a clear message and as a result there is a disconnect between heading and content.

andreas-merkel-by replied:
I wanted to motivate why we are using different clusters for different steps of the pipeline; I think that's what's special about us. I'll make this more clear.

frank-fischer-by commented on _posts/2020-03-16-dask-usage-at-blue-yonder.markdown:
```diff
@@ -0,0 +1,161 @@
+
+The number and size of the workers varies from cluster to cluster depending on the degree of parallelism
+of the computation being performed, its resource requirements, and the available timeframe for the computation.
+At the time of writing this, we are running more than 500 distinct clusters.
```
`At the time of writing` is a fixed term. You should omit the "this"

```suggestion
At the time of writing, we are running more than 500 distinct clusters.
```

frank-fischer-by commented on _posts/2020-03-16-dask-usage-at-blue-yonder.markdown:
```diff
@@ -0,0 +1,161 @@
+The number and size of the workers varies from cluster to cluster depending on the degree of parallelism
+of the computation being performed, its resource requirements, and the available timeframe for the computation.
+At the time of writing this, we are running more than 500 distinct clusters.
+They have between one and 225 workers, with worker size varying between 1GiB and 64GiB of memory.
```
You should IMHO preface, that/why we only care about memory and not about storage/cores?!

frank-fischer-by commented on _posts/2020-03-16-dask-usage-at-blue-yonder.markdown:
```diff
@@ -0,0 +1,161 @@
+of the computation being performed, its resource requirements, and the available timeframe for the computation.
+At the time of writing this, we are running more than 500 distinct clusters.
+They have between one and 225 workers, with worker size varying between 1GiB and 64GiB of memory.
+Our total memory use goes up to as much as 15TiB.
```
Total of what? One cluster or all clusters?

fjetter replied:
how many CPUs? 

Can we visualise this somehow (if there is no sane visualization, that's also fine, of course)?

fjetter commented on _posts/2020-03-16-dask-usage-at-blue-yonder.markdown:
```diff
@@ -0,0 +1,161 @@
+[Dask/Distributed](https://distributed.dask.org) and how we can leverage it to power
+our machine learning and data pipelines.
+It's 2020 now, and we are using it heavily in production for providing demand predictions,
+recommended order quantities, and price proposals to our customers.
```
Not sure if an external reader actually understands what we mean with demand predictions and oder quantiles. I think we should assume that a reader does not actually know what Blue Yonder does (and even if somebody knows BY, BY is nowadays a quite large company with many applications)

fjetter commented on _posts/2020-03-16-dask-usage-at-blue-yonder.markdown:
```diff
@@ -0,0 +1,161 @@
+as well as the predictions, order quantities, and prices themselves.
+
+The size of the data we work on varies strongly depending on the individual customer, with the largest ones
+currently amounting to about 300 GiB of compressed Parquet data per day, corresponding to about 500 GiB of
```
Where do these numbers come from?
1. 300GiB a day sounds a bit too large
2. We have usually compression rates of at least 20, i.e. 300GiB compressed would amount to about ~6TB in memory.

Maybe we can cite number of rows as well? I guess num of rows is more universal, storage size and in memory depends on so many things

andreas-merkel-by replied:
I got it from your Dask summit slides, but out of unconscious delusions of grandeur added a zero. 😂

fjetter commented on _posts/2020-03-16-dask-usage-at-blue-yonder.markdown:
```diff
@@ -0,0 +1,161 @@
+One typical use case that we have is downloading large amounts of data from a database into a
+dataset. We use Dask/Distributed here as a means to parallelize the download and the graph is
+an extremely simple one: a lot of parallel nodes downloading pieces of data into a blob and one
+reduction node writing the dataset metadata (which is basically a dictionary what data can be found in which blob).
```
Might be another opportunity to drop a link to kartothek

andreas-merkel-by replied:
I have a link to the Kartothek blog post in the section before, but I'll add one here for good measure 😄

fjetter commented on _posts/2020-03-16-dask-usage-at-blue-yonder.markdown:
```diff
@@ -0,0 +1,161 @@
+
+One typical use case that we have is downloading large amounts of data from a database into a
+dataset. We use Dask/Distributed here as a means to parallelize the download and the graph is
+an extremely simple one: a lot of parallel nodes downloading pieces of data into a blob and one
```
The technical term for "extremely simple" graph in this case would be [embarrassingly parallel](https://en.wikipedia.org/wiki/Embarrassingly_parallel)

fjetter commented on _posts/2020-03-16-dask-usage-at-blue-yonder.markdown:
```diff
@@ -0,0 +1,161 @@
+
+### Downloading data from a database to a dataset
+
+One typical use case that we have is downloading large amounts of data from a database into a
```
One important bit about these graphs would be that we also need to limit concurrency (which is why we built the semaphore. We could link https://docs.dask.org/en/latest/futures.html?highlight=semaphore#id1

fjetter commented on _posts/2020-03-16-dask-usage-at-blue-yonder.markdown:
```diff
@@ -0,0 +1,161 @@
+To improve manageability, resilience, and resource utilization, we run the Dask clusters on top
+of [Apache Mesos](http://mesos.apache.org/)/[Aurora](http://aurora.apache.org/) 
+and [Kubernetes](https://kubernetes.io/). This means every worker as well as the scheduler and client
+each run in an insolated container. Communication happens via reverse proxies to make the communication
```
```suggestion
each run in an isolated container. Communication happens via reverse proxies to make the communication
```

unless you're making a thermodynamical point I didn't get ;)

fjetter commented on _posts/2020-03-16-dask-usage-at-blue-yonder.markdown:
```diff
@@ -0,0 +1,161 @@
+The particular way we use Dask, especially running it in containers connected by reverse proxies and the fact
+that we dynamically add/remove workers from a cluster quite frequently for autoscaling has lead us to hit some
+edge cases and instabilities and given us the chance to contribute some fixes and improvements to Dask.
+For instance, we were able to 
```
By now, the list of fixes we contributed is quite long. Shall we put in more of these?

andreas-merkel-by replied:
I'm not sure if we should really put a long list in here, it could look a bit like showing off. I'd just list the top ones, not sure if I got the right ones, though.

fjetter reviewed (commented):
Overall I like the format since it gives a nice overview about what we're doing with it. We might consider putting in a Conclusion section with topics like: Are we happy/not happy. What are our dreams and visions of dask. What would we like to do with it in the future, etc.

andreas-merkel-by committed:
Add blog post about Dask usage at Blue Yonder.

andreas-merkel-by committed:
Condense and sharpen focus.
